### DISCUSSION TITLE
OIDC - Access token - Can it be refreshed ?

### DISCUSSION THREAD

**Discussion by alexandrebrg** [ory/kratos#2176] [Q&A] [answered]
Hello guys, 

I'm currently working on a project for my company, where we are using OIDC to let people register. These tokens given by the OIDC are used to display information in the user dashboard of the application (in our case, it is a GitLab, and we want to fetch user projects). These tokens are due to expire at a time, I know there is an issue to refresh them on [login](https://github.com/ory/kratos/issues/1912). But is there any way to refresh the access token from Kratos without having to ask the user to log again ? 

Another solution I'm thinking of (but it will lead into security concerns), it is possible to update (with our data) of the user OIDC provider data ? 

Thanks!

**Comment by vinckr** [comment, marked as answer]
Hello @AlexandreBrg 
In general I think it is the best option to refresh them on login, but if you want to continuously fetch user projects I can see how that presents a problem.
How long is the TTL on the gitlab tokens?
When do you think would be a good time to refresh them or what method can be used?
It am not sure of the implications, but as you mentioned it sounds to me as this could be a security concern, it probably depends on the details of what gets refreshed, when & how.

**Reply by alexandrebrg** [reply]
The main idea of it was to create a OIDC refresh flow in order to refresh a specific provider tokens. This flow can be called at anytime by the authenticated user, but in order to know when to call this route we must know when the token expires, and it's an information we don't have except from Admin API. And as this flow would be triggered automatically, I guess it would an API only flow. This implementation means a big development and a lot of security concerns with it, I guess. 

That's why having a refresh only when the user logs in isn't much of a bad thing, and as GitLab Access token has a lifespan of 2 hours, a user session can be defined with the same lifespan, so it never expires. Except if you got any other idea, I think it's the best trade-off to our use-case (and to people with similar use-case). I'll take a look if other identity management platforms have the same problem and how they handle it.

**Reply by vinckr** [reply]
I agree, having the refresh on login seems to be the most straightforward and I can't think of a case right now where this would lead to an issue. 
Let us know if you get around to finding out how others handle it :)

**Reply by aeneasr** [reply]
To extedn @vinckr answer a bit - we currently do not refresh the tokens because it is a bit tricky to do so reliably. For example there can be race conditions which could lead to the tokens being revoked if they get reused on accident in a racy behavior. That's why it's probably the best solution to copy the tokens into your system and use them there in the way you seem fit.

It would still be possible to add this to Kratos though of course, but it is a bit difficult to implement reliably and at scale I think.

**Reply by alexandrebrg** [reply]
Having an OIDC refresh flow is clearly not reliable as it could also lead to some kind of DOS where the authorization provider could rate limit requests. As told in #1912, it should be soon implemented, for now, we cannot count on it.  

In the meantime, as my company needs it, I designed  an external service (straightforward one) which receive a login notification based on Kratos hooks. Still, it cannot be developed because of a problem: as defined in the [RFC](https://datatracker.ietf.org/doc/html/rfc6749#section-4.1), old tokens are revoked when issuing new ones. On top of that, when using hooks, kratos only gives information about the used flow or identity, without OIDC information (like new tokens). I was thinking that, even though there are security concerns for it, couldn't these tokens be exposed in login/register hooks ? JsonNet is used to control data exposed to hooks, so security issues can be limited. What are your thought guys about it ?

**Reply by aeneasr** [reply]
> I was thinking that, even though there are security concerns for it, couldn't these tokens be exposed in login/register hooks

Yes, we can absolutely introduce that! Please feel free to create a PR with this change or an issue!

**Reply by vinckr** [reply]
Hey @AlexandreBrg, how are you doing?
just checking in if your problem still persists. 
I think there will be some progress in the linked issue soon and feel free to open one for your particular use case.
I will mark this as answered for now, but feel free to come back to the discussion any time.
Thanks a ton :bee:

**Reply by alexandrebrg** [reply]
👋  @vinckr, doing well, I hope it is the case for you as well! 

I'm not sure if the problem persists, AFAIK we managed to solve the issue with a workaround. Still, really enjoy seeing progress on this topic, it was a major pain point. 😄
